tmt/base.py

```diff
@@ -237,16 +237,206 @@ def _validate_fmf_node(self, node: fmf.Tree, logger: tmt.utils.Common,
+            for name, value in base.__dict__.get('__annotations__', {}).items():
+                # Skip special fields that are not keys.
+                if name == 'KEYS_SHOW_ORDER':
+                    continue
+
+                yield (name, value)
+
+        yield from cls.__dict__.get('__annotations__', {}).items()
+
+    @classmethod
+    def iter_key_names(cls) -> Generator[str, None, None]:
+        """
+        Iterate over key names.
+
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+
+        Yields:
+            key names.
+        """
+
+        for keyname, _ in cls._iter_key_annotations():
+            yield keyname
+
+    def iter_keys(self) -> Generator[Tuple[str, Any], None, None]:
+        """
+        Iterate over keys and their values.
+
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+
+        Yields:
+            pairs of key name and its value.
+        """
+
+        for keyname in self.iter_key_names():
+            yield (keyname, getattr(self, keyname))
+
+    # TODO: exists for backward compatibility for the transition period. Once full
+    # type annotations land, there should be no need for extra _keys attribute.
+    @classmethod
+    def _keys(cls) -> List[str]:
+        """ Return a list of names of object's keys. """
+
+        return list(cls.iter_key_names())
+
+    def _load_keys(
+            self,
+            node: fmf.Tree,
+            logger: tmt.utils.Common) -> None:
+        """ Extract values for class-level attributes, and verify they match declared types. """
+
+        for keyname, keytype in self._iter_key_annotations():
+            key_address = f'{node.name}{keyname}'
+
+            shift = 2
+
+            logger.debug('key', key_address, shift=shift)
+
+            shift += 1
+
+            logger.debug('desired type', str(keytype), shift=shift)
+
+            value: Any = None
+
+            logger.debug('dict', self.__dict__, shift=shift)
+
+            if hasattr(self, keyname):
+                # If the key exists as instance's attribute already, it is because it's been
+                # declared with a default value, and the attribute now holds said default value.
+                default_value = getattr(self, keyname)
+
+                # If the default value is a mutable container, we cannot use it directly.
+                # Should we do so, the very same default value would be assigned to multiple
+                # instances/attributes instead of each instance having its own distinct container.
+                if isinstance(default_value, (list, dict)):
+                    logger.debug('detected mutable default')
```

Comment:
Probably use `level=3` for these detailed messages?

Reply:
Or, perhaps, even `level=4`?

Reply:
Sure, it's not critical.

Reply:
Fixed, `level=4` it is.